ytxp-plutarch/src/Cardano/YTxP/Control/Yielding.hs

```diff
@@ -1,119 +1,186 @@
   }
 
 newtype PYieldingRedeemer (s :: S)
   = PYieldingRedeemer
       ( Term
           s
           ( PDataRecord
-              '[ "yieldListIndex" ':= PYieldListIndex
-               , "yieldListScriptToYieldIndex" ':= PYieldListScriptToYieldIndex
-               , "yieldListRefInputIndex" ':= PYieldListRefInputIndex
+              '[ "yrAuthorisedScriptIndex" ':= PAuthorisedScriptIndex
+               , "yrAuthorisedScriptProofIndex" ':= PAuthorisedScriptProofIndex
                ]
           )
       )
   deriving stock (Generic)
   deriving anyclass (PlutusType, PIsData, PDataFields)
 
 instance DerivePlutusType PYieldingRedeemer where
+  -- TODO/QUESTION: why arent we using PlutusTypeNewtype here?
   type DPTStrat _ = PlutusTypeData
 
 instance PTryFrom PData (PAsData PYieldingRedeemer)
 
+--------------------------------------------------------------------------------
+
+-- * Hashes
+
+------------------------------------------------------------
+
+-- ** CustomScriptHash
+
+-- We use this because the plutus-ledger-api ScriptHash isn't
+-- type safe.
+
+{- | We use this `CustomScriptHash` instead of `ScriptHash` in
+order to ensure that the hash is of length 28.
+-}
+newtype CustomScriptHash = CustomScriptHash {getCustomScriptHash :: Builtins.BuiltinByteString}
+  deriving stock
+    ( Show
+    , Eq
+    )
+
+instance PlutusTx.UnsafeFromData CustomScriptHash where
+  {-# INLINEABLE unsafeFromBuiltinData #-}
+  unsafeFromBuiltinData b =
+    let !args = BI.snd $ BI.unsafeDataAsConstr b
+        scriptHash = BI.unsafeDataAsB (BI.head args)
+     in CustomScriptHash scriptHash
+
+instance PlutusTx.ToData CustomScriptHash where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (CustomScriptHash scriptHash) = PlutusTx.toBuiltinData scriptHash
+
+instance PlutusTx.FromData CustomScriptHash where
+  {-# INLINEABLE fromBuiltinData #-}
+  fromBuiltinData b = do
+    scriptHash <- PlutusTx.fromBuiltinData b
+    guard (Builtins.lengthOfByteString scriptHash == 28)
+    pure $ tryMkCustomScriptHash scriptHash
+
+{- | Note(Nigel): This will likely not compile under `plutus-tx`
+due to the use of `error` from the Haskell `Prelude`.
+We use `error` from Prelude here as using `traceError` doesn't give back the error message.
+See the following issue for more details: https://github.com/IntersectMBO/plutus/issues/3003
+-}
+{-# INLINEABLE tryMkCustomScriptHash #-}
+tryMkCustomScriptHash :: Builtins.BuiltinByteString -> CustomScriptHash
+tryMkCustomScriptHash scriptHash
+  | Builtins.lengthOfByteString scriptHash /= 28 =
+      error "tryMkCustomScriptHash: ScriptHash must have length 28"
+  | otherwise = CustomScriptHash scriptHash
+
+------------------------------------------------------------
+
+-- ** AuthorisedScriptHash
+
+-- | A single hash that a yielding script can yield to
+newtype AuthorisedScriptHash = AuthorisedScriptHash CustomScriptHash
```

Comment:
We don't need this `AuthorisedScriptHash` type anymore. Before we were encoding this information in the yield list datum, so that is why we wanted to wrap this into the `CustomScriptHash`, now we get the scripthash from the reference input at the pointed index (checking for the ST token to ensure the ref input is "valid"). The scripthash will already be an onchain well-formed PScriptHash so no need to do any extra wrapping